File: src/solver/subroutines/solveYellowCorners.ts
import { cubesShareFace, positionMap } from '../../utils'
import RubiksCubeSolver from '../RubiksCubeSolver'

export default async function solveYellowCorners(solver: RubiksCubeSolver) {
  /////////////////////////////////
  // Solve yellow top corner cubes
  /////////////////////////////////
  const backLeft = solver.cubeMap.get(positionMap[1])
  const backEdge = solver.cubeMap.get(positionMap[2])
  const leftEdge = solver.cubeMap.get(positionMap[4])
  if (
    backLeft?.hasFace('Y') &&
    !(
      backLeft?.orientation.top === 'Y' &&
      cubesShareFace('left', backLeft, leftEdge) &&
      cubesShareFace('back', backLeft, backEdge)
    )
  ) {
    await solver.do('XCCW')
    await solveFrontLeftCorner(solver)
    return
  }

  const backRight = solver.cubeMap.get(positionMap[3])
  const rightEdge = solver.cubeMap.get(positionMap[6])
  if (
    backRight?.hasFace('Y') &&
    !(
      backRight?.orientation.top === 'Y' &&
      cubesShareFace('right', backRight, rightEdge) &&
      cubesShareFace('back', backRight, backEdge)
    )
  ) {
    await solver.do('XCW')
    await solveFrontRightCorner(solver)
    return
  }

  const frontLeft = solver.cubeMap.get(positionMap[7])
  const frontEdge = solver.cubeMap.get(positionMap[8])
  if (
    frontLeft?.hasFace('Y') &&
    !(
      frontLeft?.orientation.top === 'Y' &&
      cubesShareFace('left', frontLeft, leftEdge) &&
      cubesShareFace('front', frontLeft, frontEdge)
    )
  ) {
    await solveFrontLeftCorner(solver)
    return
  }

  const frontRight = solver.cubeMap.get(positionMap[9])
  if (
    frontRight?.hasFace('Y') &&
    !(
      frontRight?.orientation.top === 'Y' &&
      cubesShareFace('right', frontRight, rightEdge) &&
      cubesShareFace('front', frontRight, frontEdge)
    )
  ) {
    await solveFrontRightCorner(solver)
    return
  }

  ////////////////////////////////////
  // Solve yellow bottom corner cubes
  ////////////////////////////////////
  const bottomBackLeft = solver.cubeMap.get(positionMap[19])
  if (bottomBackLeft?.hasFace('Y')) {
    if (bottomBackLeft.orientation.bottom === 'Y') {
      await solver.do('XCCW')
      await solveBottomLeftFaceDown(solver)
    } else if (bottomBackLeft.orientation.left === 'Y') {
      await solver.do('rotateBottomCCW')
      await solveBottomLeft(solver)
    } else {
      await solver.do('rotateBottomCCW', 'rotateBottomCCW')
      await solveBottomRight(solver)
    }
    return
  }

  const bottomBackRight = solver.cubeMap.get(positionMap[21])
  if (bottomBackRight?.hasFace('Y')) {
    if (bottomBackRight.orientation.bottom === 'Y') {
      await solver.do('XCW')
      await solveBottomRightFaceDown(solver)
    } else if (bottomBackRight.orientation.right === 'Y') {
      await solver.do('rotateBottomCW')
      await solveBottomRight(solver)
    } else {
      await solver.do('rotateBottomCW', 'rotateBottomCW')
      await solveBottomLeft(solver)
    }
    return
  }

  const bottomFrontLeft = solver.cubeMap.get(positionMap[25])
  if (bottomFrontLeft?.hasFace('Y')) {
    if (bottomFrontLeft.orientation.bottom === 'Y') {
      await solveBottomLeftFaceDown(solver)
    } else if (bottomFrontLeft.orientation.left === 'Y') {
      await solver.do('rotateBottomCCW')
      await solveBottomRight(solver)
    } else {
      await solveBottomLeft(solver)
    }
    return
  }

  const bottomFrontRight = solver.cubeMap.get(positionMap[27])
  if (bottomFrontRight?.hasFace('Y')) {
    if (bottomFrontRight.orientation.bottom === 'Y') {
      await solveBottomRightFaceDown(solver)
    } else if (bottomFrontRight.orientation.right === 'Y') {
      await solver.do('rotateBottomCW')
      await solveBottomLeft(solver)
    } else {
      await solveBottomRight(solver)
    }
    return
  }
}

////////////////////////////////////
// Solve for specific cubes
////////////////////////////////////
async function solveFrontRightCorner(solver: RubiksCubeSolver) {
  const unsolvedCube = solver.cubeMap.get(positionMap[9])
  if (unsolvedCube?.orientation.front === 'Y') {
    await solver.do('rotateFrontCW', 'rotateBottomCCW', 'rotateFrontCCW', 'XCW')
    await solveBottomRight(solver)
  } else if (unsolvedCube?.orientation.top === 'Y') {
    await solver.do('rotateRightCCW', 'rotateBottomCW', 'rotateRightCW', 'XCCW')
    await solveBottomRight(solver)
  } else {
    await solver.do('rotateRightCCW', 'rotateBottomCW', 'rotateRightCW')
    await solveBottomLeft(solver)
  }
  return
}

async function solveFrontLeftCorner(solver: RubiksCubeSolver) {
  const unsolvedCube = solver.cubeMap.get(positionMap[7])
  if (unsolvedCube?.orientation.front === 'Y') {
    await solver.do('rotateFrontCCW', 'rotateBottomCW', 'rotateFrontCW', 'XCCW')
    await solveBottomLeft(solver)
  } else if (unsolvedCube?.orientation.top === 'Y') {
    await solver.do('rotateLeftCCW', 'rotateBottomCCW', 'rotateLeftCW', 'XCW')
    await solveBottomLeft(solver)
  } else {
    await solver.do('rotateLeftCCW', 'rotateBottomCCW', 'rotateLeftCW')
    await solveBottomRight(solver)
  }
  return
}

async function solveBottomLeftFaceDown(solver: RubiksCubeSolver) {
  let unsolvedWorkingCube = false
  do {
    const topLeftCorner = solver.cubeMap.get(positionMap[7])
    const topLeftEdge = solver.cubeMap.get(positionMap[4])
    const topFrontEdge = solver.cubeMap.get(positionMap[8])
    if (
      topLeftCorner?.orientation.top !== 'Y' ||
      !cubesShareFace('left', topLeftCorner, topLeftEdge) ||
      !cubesShareFace('front', topLeftCorner, topFrontEdge)
    ) {
      unsolvedWorkingCube = true
    } else {
      await solver.do('rotateBottomCW', 'XCCW')
    }
  } while (!unsolvedWorkingCube)

  await solver.do(
    'rotateBottomCCW',
    'rotateLeftCCW',
    'rotateBottomCW',
    'rotateLeftCW',
    'rotateFrontCCW',
    'rotateBottomCW',
    'rotateFrontCW',
    'rotateBottomCCW'
  )
  await solveBottomLeft(solver)
  return
}

async function solveBottomRightFaceDown(solver: RubiksCubeSolver) {
  let unsolvedWorkingCube = false
  do {
    const topRightCorner = solver.cubeMap.get(positionMap[9])
    const topRightEdge = solver.cubeMap.get(positionMap[6])
    const topFrontEdge = solver.cubeMap.get(positionMap[8])
    if (
      topRightCorner?.orientation.top !== 'Y' ||
      !cubesShareFace('right', topRightCorner, topRightEdge) ||
      !cubesShareFace('front', topRightCorner, topFrontEdge)
    ) {
      unsolvedWorkingCube = true
    } else {
      await solver.do('rotateBottomCCW', 'XCW')
    }
  } while (!unsolvedWorkingCube)

  await solver.do(
    'rotateBottomCW',
    'rotateRightCCW',
    'rotateBottomCCW',
    'rotateRightCW',
    'rotateFrontCW',
    'rotateBottomCCW',
    'rotateFrontCCW',
    'rotateBottomCW'
  )
  await solveBottomRight(solver)
  return
}

async function solveBottomLeft(solver: RubiksCubeSolver) {
  const cube = solver.cubeMap.get(positionMap[25])
  while (!cubesShareFace('left', cube, solver.cubeMap.get(positionMap[13]))) {
    await solver.do('rotateBottomCW', 'XCCW')
  }
  await solver.do('rotateBottomCCW', 'rotateLeftCCW', 'rotateBottomCW', 'rotateLeftCW')
  return
}

async function solveBottomRight(solver: RubiksCubeSolver) {
  const cube = solver.cubeMap.get(positionMap[27])
  while (!cubesShareFace('right', cube, solver.cubeMap.get(positionMap[15]))) {
    await solver.do('rotateBottomCCW', 'XCW')
  }
  await solver.do('rotateBottomCW', 'rotateRightCCW', 'rotateBottomCCW', 'rotateRightCW')
  return
}

File: android/core/src/main/kotlin/com/mikeb/simplepuzzlecube/solver/subroutines/solveYellowCorners.kt
// Port of ../src/solver/subroutines/solveYellowCorners.ts.
package com.mikeb.simplepuzzlecube.solver.subroutines

import com.mikeb.simplepuzzlecube.cubesShareFace
import com.mikeb.simplepuzzlecube.engine.Face
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateBottomCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateBottomCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateFrontCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateFrontCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateLeftCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateLeftCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateRightCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateRightCW
import com.mikeb.simplepuzzlecube.engine.OrientationKey
import com.mikeb.simplepuzzlecube.engine.Rotation.XCCW
import com.mikeb.simplepuzzlecube.engine.Rotation.XCW
import com.mikeb.simplepuzzlecube.solver.RubiksCubeSolver

suspend fun solveYellowCorners(solver: RubiksCubeSolver) {
    /////////////////////////////////
    // Solve yellow top corner cubes
    /////////////////////////////////
    val backLeft = solver.fetchPosition(1)
    val backEdge = solver.fetchPosition(2)
    val leftEdge = solver.fetchPosition(4)
    if (
        backLeft?.hasFace(Face.Y) == true &&
        !(
            backLeft.orientation.top == Face.Y &&
                cubesShareFace(OrientationKey.left, backLeft, leftEdge) &&
                cubesShareFace(OrientationKey.back, backLeft, backEdge)
            )
    ) {
        solver.`do`(XCCW)
        solveFrontLeftCorner(solver)
        return
    }

    val backRight = solver.fetchPosition(3)
    val rightEdge = solver.fetchPosition(6)
    if (
        backRight?.hasFace(Face.Y) == true &&
        !(
            backRight.orientation.top == Face.Y &&
                cubesShareFace(OrientationKey.right, backRight, rightEdge) &&
                cubesShareFace(OrientationKey.back, backRight, backEdge)
            )
    ) {
        solver.`do`(XCW)
        solveFrontRightCorner(solver)
        return
    }

    val frontLeft = solver.fetchPosition(7)
    val frontEdge = solver.fetchPosition(8)
    if (
        frontLeft?.hasFace(Face.Y) == true &&
        !(
            frontLeft.orientation.top == Face.Y &&
                cubesShareFace(OrientationKey.left, frontLeft, leftEdge) &&
                cubesShareFace(OrientationKey.front, frontLeft, frontEdge)
            )
    ) {
        solveFrontLeftCorner(solver)
        return
    }

    val frontRight = solver.fetchPosition(9)
    if (
        frontRight?.hasFace(Face.Y) == true &&
        !(
            frontRight.orientation.top == Face.Y &&
                cubesShareFace(OrientationKey.right, frontRight, rightEdge) &&
                cubesShareFace(OrientationKey.front, frontRight, frontEdge)
            )
    ) {
        solveFrontRightCorner(solver)
        return
    }

    ////////////////////////////////////
    // Solve yellow bottom corner cubes
    ////////////////////////////////////
    val bottomBackLeft = solver.fetchPosition(19)
    if (bottomBackLeft?.hasFace(Face.Y) == true) {
        if (bottomBackLeft.orientation.bottom == Face.Y) {
            solver.`do`(XCCW)
            solveBottomLeftFaceDown(solver)
        } else if (bottomBackLeft.orientation.left == Face.Y) {
            solver.`do`(rotateBottomCCW)
            solveBottomLeft(solver)
        } else {
            solver.`do`(rotateBottomCCW, rotateBottomCCW)
            solveBottomRight(solver)
        }
        return
    }

    val bottomBackRight = solver.fetchPosition(21)
    if (bottomBackRight?.hasFace(Face.Y) == true) {
        if (bottomBackRight.orientation.bottom == Face.Y) {
            solver.`do`(XCW)
            solveBottomRightFaceDown(solver)
        } else if (bottomBackRight.orientation.right == Face.Y) {
            solver.`do`(rotateBottomCW)
            solveBottomRight(solver)
        } else {
            solver.`do`(rotateBottomCW, rotateBottomCW)
            solveBottomLeft(solver)
        }
        return
    }

    val bottomFrontLeft = solver.fetchPosition(25)
    if (bottomFrontLeft?.hasFace(Face.Y) == true) {
        if (bottomFrontLeft.orientation.bottom == Face.Y) {
            solveBottomLeftFaceDown(solver)
        } else if (bottomFrontLeft.orientation.left == Face.Y) {
            solver.`do`(rotateBottomCCW)
            solveBottomRight(solver)
        } else {
            solveBottomLeft(solver)
        }
        return
    }

    val bottomFrontRight = solver.fetchPosition(27)
    if (bottomFrontRight?.hasFace(Face.Y) == true) {
        if (bottomFrontRight.orientation.bottom == Face.Y) {
            solveBottomRightFaceDown(solver)
        } else if (bottomFrontRight.orientation.right == Face.Y) {
            solver.`do`(rotateBottomCW)
            solveBottomLeft(solver)
        } else {
            solveBottomRight(solver)
        }
        return
    }
}

////////////////////////////////////
// Solve for specific cubes
////////////////////////////////////
private suspend fun solveFrontRightCorner(solver: RubiksCubeSolver) {
    val unsolvedCube = solver.fetchPosition(9)
    if (unsolvedCube?.orientation?.front == Face.Y) {
        solver.`do`(rotateFrontCW, rotateBottomCCW, rotateFrontCCW, XCW)
        solveBottomRight(solver)
    } else if (unsolvedCube?.orientation?.top == Face.Y) {
        solver.`do`(rotateRightCCW, rotateBottomCW, rotateRightCW, XCCW)
        solveBottomRight(solver)
    } else {
        solver.`do`(rotateRightCCW, rotateBottomCW, rotateRightCW)
        solveBottomLeft(solver)
    }
    return
}

private suspend fun solveFrontLeftCorner(solver: RubiksCubeSolver) {
    val unsolvedCube = solver.fetchPosition(7)
    if (unsolvedCube?.orientation?.front == Face.Y) {
        solver.`do`(rotateFrontCCW, rotateBottomCW, rotateFrontCW, XCCW)
        solveBottomLeft(solver)
    } else if (unsolvedCube?.orientation?.top == Face.Y) {
        solver.`do`(rotateLeftCCW, rotateBottomCCW, rotateLeftCW, XCW)
        solveBottomLeft(solver)
    } else {
        solver.`do`(rotateLeftCCW, rotateBottomCCW, rotateLeftCW)
        solveBottomRight(solver)
    }
    return
}

private suspend fun solveBottomLeftFaceDown(solver: RubiksCubeSolver) {
    var unsolvedWorkingCube = false
    do {
        val topLeftCorner = solver.fetchPosition(7)
        val topLeftEdge = solver.fetchPosition(4)
        val topFrontEdge = solver.fetchPosition(8)
        if (
            topLeftCorner?.orientation?.top != Face.Y ||
            !cubesShareFace(OrientationKey.left, topLeftCorner, topLeftEdge) ||
            !cubesShareFace(OrientationKey.front, topLeftCorner, topFrontEdge)
        ) {
            unsolvedWorkingCube = true
        } else {
            solver.`do`(rotateBottomCW, XCCW)
        }
    } while (!unsolvedWorkingCube)

    solver.`do`(
        rotateBottomCCW,
        rotateLeftCCW,
        rotateBottomCW,
        rotateLeftCW,
        rotateFrontCCW,
        rotateBottomCW,
        rotateFrontCW,
        rotateBottomCCW
    )
    solveBottomLeft(solver)
    return
}

private suspend fun solveBottomRightFaceDown(solver: RubiksCubeSolver) {
    var unsolvedWorkingCube = false
    do {
        val topRightCorner = solver.fetchPosition(9)
        val topRightEdge = solver.fetchPosition(6)
        val topFrontEdge = solver.fetchPosition(8)
        if (
            topRightCorner?.orientation?.top != Face.Y ||
            !cubesShareFace(OrientationKey.right, topRightCorner, topRightEdge) ||
            !cubesShareFace(OrientationKey.front, topRightCorner, topFrontEdge)
        ) {
            unsolvedWorkingCube = true
        } else {
            solver.`do`(rotateBottomCCW, XCW)
        }
    } while (!unsolvedWorkingCube)

    solver.`do`(
        rotateBottomCW,
        rotateRightCCW,
        rotateBottomCCW,
        rotateRightCW,
        rotateFrontCW,
        rotateBottomCCW,
        rotateFrontCCW,
        rotateBottomCW
    )
    solveBottomRight(solver)
    return
}

private suspend fun solveBottomLeft(solver: RubiksCubeSolver) {
    val cube = solver.fetchPosition(25)
    while (!cubesShareFace(OrientationKey.left, cube, solver.fetchPosition(13))) {
        solver.`do`(rotateBottomCW, XCCW)
    }
    solver.`do`(rotateBottomCCW, rotateLeftCCW, rotateBottomCW, rotateLeftCW)
    return
}

private suspend fun solveBottomRight(solver: RubiksCubeSolver) {
    val cube = solver.fetchPosition(27)
    while (!cubesShareFace(OrientationKey.right, cube, solver.fetchPosition(15))) {
        solver.`do`(rotateBottomCCW, XCW)
    }
    solver.`do`(rotateBottomCW, rotateRightCCW, rotateBottomCCW, rotateRightCW)
    return
}
